File: Python/BOJ/9012.py
import sys

def solution():
    N = int(sys.stdin.readline())
    answer = []
    for i in range(N):
        parenthesis = sys.stdin.readline().strip()
        stack = []
        for p in parenthesis:
            stack.append(p)
            if len(stack) > 1 and stack[-2:] == ['(', ')']:
                del stack[-2:]
        #print(stack)
        if (len(stack) != 0):
            print("NO")
        else:
            print("YES")

solution()

File: C++/BOJ/9012.cpp
#include<vector>
#include<iostream>
#include<string>
using namespace std;

int main()
{
    int N;
    string  parenthesis = "";
    cin>> N;
    for (int i = 0; i < N; i++)
    {
        vector<char> stack;
        cin>>parenthesis;
        for (auto p : parenthesis)
        {
            stack.push_back(p);
            vector<char>::reverse_iterator r_it = stack.rbegin();
            if (stack.size() > 1)
            {
                if (*r_it == ')' && *(++r_it) == '(')
                {
                    stack.pop_back();
                    stack.pop_back();
                }
            }
        }
        if (stack.size() != 0)
            cout<<"NO\n";
        else
            cout<<"YES\n";
    }
}